### DISCUSSION TITLE
HS400 eMMC mode on the icicle kit

### DISCUSSION THREAD

**Discussion by fcuzzocrea** [polarfire-soc/polarfire-soc-documentation#22] [Bare metal embedded software] [answered]
Hi!

According to ICICLE Kit Documentation (DS60001679B - ICICLE_Kit_User_Guide_VB) the eMMC of the ICICLE Kit (Sandisk SDINBDG4-8G) should support HS400 bus speed mode (coupled with 200 MHz clock speed). However, when setting this either in a baremental application or in the HSS, the MSS_MMC_init function fails to initialize the eMMC. I spotted the issue when trying to configure the eMMC on a custom design and realized that HS400 wasn't working, so I decided to try on the ICICLE Kit and had the same behavior.

Should this be considered as a bug of the BSP or is a not supported mode?

**Comment by padmaraob** [comment]
Hi @fcuzzocrea, It is supporting HS400 mode, will look into the issue and comeback to you

**Comment by padmaraob** [comment, marked as answer]
The PolarFire SoC MSS Configurator is setting the eMMC mssio output drive 8mA by default if it change to 10mA then all eMMC bus speed modes work fine including HS400. The Configurator generates XML file that can be used in the project.

Or

Change Bank 4 I/Os value 0x0928 to 0x0930 in the existing XML file like below

MSSIO_BANK4_IO_CFG_0_1_CR:

```
> <register address="0x00000234" description="IO electrical configuration for MSSIO pad" name="MSSIO_BANK4_IO_CFG_0_1_CR">
>             <field Type="RW" name="IO_CFG_0" offset="0" width="16">0x0930</field>
>             <field Type="RW" name="IO_CFG_1" offset="16" width="16">0x0930</field>

```
also change
 
MSSIO_BANK4_IO_CFG_2_3_CR
MSSIO_BANK4_IO_CFG_4_5_CR
MSSIO_BANK4_IO_CFG_6_7_CR
MSSIO_BANK4_IO_CFG_8_9_CR
MSSIO_BANK4_IO_CFG_10_11_CR
MSSIO_BANK4_IO_CFG_12_13_CR

HSS XML : https://github.com/polarfire-soc/hart-software-services/blob/master/boards/mpfs-icicle-kit-es/soc_fpga_design/xml/ICICLE_MSS_mss_cfg.xml

Bare Metal : https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/blob/main/driver-examples/mss/mss-mmc/mpfs-emmc-sd-write-read/src/boards/icicle-kit-es/fpga_design/design_description/ICICLE_MSS_eMMC_cfg.xml

Once modified XML file then build the example project(HSS or Baremetal)

**Comment by fcuzzocrea** [comment]
Hi!!

I just tried what you suggested in the previous comment on my icicle kit, using HSS.

I found out that the HSS half-ignores what is generated with the polarfire soc configurator, as the #define are hardcoded here: https://github.com/polarfire-soc/hart-software-services/blob/01e82e3e05b5b00a22920eba811d95ed885ff16d/services/mmc/mmc_api.c#L66

I changed also the values of those defines to reflect what the configurator tool was setting (0x09300930UL), however I had a very weird behavior.

after HSS boots, my icicle kit first prints this:

`[10.115690] getBootImageFromMMC_(): copyBootImageToDDR_() failed`

then, after some seconds, this is spammed continuously:

`[374.144994] boot_download_chunks_handler(): boot_service(u54_1)::Skipping chunk 0xa01233f8 due to ownership -1`

Do you have any hints to decode this behaviour?

Also, I noticed that on HSS those registers are manually set as part of the initialization stage:

https://github.com/polarfire-soc/hart-software-services/blob/01e82e3e05b5b00a22920eba811d95ed885ff16d/services/mmc/mmc_api.c#L132

I thought that, according to bsp documentation, only the configuration structure should be set https://github.com/polarfire-soc/platform/blob/48c848c6ce788e517b67aab356c1a385799b63a8/drivers/mss/mss_mmc/mss_mmc.h#L372.

So, should I manually set those registers as part of initialization flow?

Thank you!!

EDIT: 

Instead of changing the values using the Configurator, I manually edited the provided XML of the HSS. Now the HSS is booting fine, but still it seems that eMMC is not working properly. In particular:

 * if I expose the eMMC as mass storage with the usbdmsc command, the HSS softreboots
 * the onboard linux does not start, just sits at `Starting kernel ...`
 
 This is the output of my linux box dmesg when that happens
 
```
[435772.164669] usb 3-3: new high-speed USB device number 95 using xhci_hcd
[435772.317429] usb 3-3: New USB device found, idVendor=1514, idProduct=0001, bcdDevice=30.00
[435772.317438] usb 3-3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[435772.317441] usb 3-3: Product: PolarFireSoc-FlashDrive
[435772.317444] usb 3-3: Manufacturer: Microchip Inc
[435772.317446] usb 3-3: SerialNumber: DF68A16A0C0A324CD0540B61
[435772.319400] usb-storage 3-3:1.0: USB Mass Storage device detected
[435772.320090] scsi host0: usb-storage 3-3:1.0
[435773.325234] scsi 0:0:0:0: Direct-Access     MSCC     PolarFireSoC_msd 1234 PQ: 0 ANSI: 4
[435773.325450] sd 0:0:0:0: Attached scsi generic sg0 type 0
[435773.331178] sd 0:0:0:0: [sda] 15273600 512-byte logical blocks: (7.82 GB/7.28 GiB)
[435773.339879] sd 0:0:0:0: [sda] Write Protect is off
[435773.339883] sd 0:0:0:0: [sda] Mode Sense: 00 00 00 00
[435773.341088] sd 0:0:0:0: [sda] Asking for cache data failed
[435773.341091] sd 0:0:0:0: [sda] Assuming drive cache: write through
```

**Comment by padmaraob** [comment]
Hi @fcuzzocrea,
Can you change eMMC hard coded value **0x09280928UL** to **0x09400940UL** in the mmc_api.c? 
https://github.com/polarfire-soc/hart-software-services/blob/master/services/mmc/mmc_api.c#L66

#define LIBERO_SETTING_MSSIO_BANK4_IO_CFG_0_1_CR_eMMC	0x09400940UL
#define LIBERO_SETTING_MSSIO_BANK4_IO_CFG_2_3_CR_eMMC	0x09400940UL
#define LIBERO_SETTING_MSSIO_BANK4_IO_CFG_4_5_CR_eMMC	0x09400940UL
#define LIBERO_SETTING_MSSIO_BANK4_IO_CFG_6_7_CR_eMMC	0x09400940UL
#define LIBERO_SETTING_MSSIO_BANK4_IO_CFG_8_9_CR_eMMC	0x09400940UL
#define LIBERO_SETTING_MSSIO_BANK4_IO_CFG_10_11_CR_eMMC	0x09400940UL
#define LIBERO_SETTING_MSSIO_BANK4_IO_CFG_12_13_CR_eMMC	0x09400940UL

change mode  **.bus_speed_mode = MSS_MMC_MODE_HS400,**
https://github.com/polarfire-soc/hart-software-services/blob/master/services/mmc/mmc_api.c#L162

use 2022.02 release (HSS, icicle-kit-reference-design, meta-polarfire-soc-yocto-bsp)

**Comment by hughbreslin** [comment]
Hey @fcuzzocrea are you still seeing this issue? The latest MMC driver has updates for HS400 support and is available in the [v2022.08](https://github.com/polarfire-soc/platform/releases/tag/2022.08) release

**Reply by fcuzzocrea** [reply]
Thanks, wem trying to use the latest eMMC driver, but we are getting some issues in getting the driver initialized correctly.

In particular, to initialize the driver the flow which we are following is:

- we turn on the clock and remove the peripheral from soft reset 
```
   SYSREG->SUBBLK_CLOCK_CR |= SUBBLK_CLOCK_CR_MMC_MASK;
   SYSREG->SOFT_RESET_CR &= (uint32_t) ~(SUBBLK_CLOCK_CR_MMC_MASK);
```

- we init the eMMC driver as follows:
```
   void init_emmc()
{
    static mss_mmc_cfg_t emmcConfig;

    emmcConfig.clk_rate = MSS_MMC_CLOCK_200MHZ;
    emmcConfig.card_type = MSS_MMC_CARD_TYPE_MMC;
    emmcConfig.bus_speed_mode = MSS_MMC_MODE_HS200;
    emmcConfig.data_bus_width = MSS_MMC_DATA_WIDTH_8BIT;
    emmcConfig.bus_voltage = MSS_MMC_1_8V_BUS_VOLTAGE;

    MSS_MMC_init(&emmcConfig);

    /* DMA init for eMMC */
    MSS_MPU_configure(
            MSS_MPU_MMC, MSS_MPU_PMP_REGION3, LIM_BASE_ADDRESS, LIM_SIZE,
            MPU_MODE_READ_ACCESS | MPU_MODE_WRITE_ACCESS | MPU_MODE_EXEC_ACCESS,
            MSS_MPU_AM_NAPOT, 0u);
}
```

- The initialization always stucks here: 
    https://github.com/polarfire-soc/platform/blob/7071e72f6f140547e7db1a2e8d2445dce1f9d4ba/drivers/mss/mss_mmc/mss_mmc.c#L5453
  ret_status is MSS_MMC_ERR_INTERRUPT

Are we missing something in the initialization?

Thanks!

**Reply by hughbreslin** [reply]
Hmm no I don't think so, are you using the latest XML from the reference design? Theres a copy [here](https://github.com/polarfire-soc/icicle-kit-reference-design/blob/master/XML/MPFS_Icicle_MSS_Linux_cfg.xml). This has a small change to the drive strength for the MMC I/Os which may help with the issue you're having.

**Reply by fcuzzocrea** [reply]
Hi, thanks for your reply. We tried the XML you provided but didn't helped.
The issue is present only when we build our custom software, the HSS does not seem to have this issue. We are asking if some other register needs to be set, because for example the HSS is setting the following in SYSREGS: https://github.com/polarfire-soc/hart-software-services/blob/01cd12714499d5c4d5bfd9b48a75621fe9a2421b/services/mmc/mmc_api.c#L135

For the record, we are doing the same in our custom build but didn't helped. Our custom build is running on the same board that the HSS has been run, so it seems to be a software configuration issue.

Thanks!

**Reply by hughbreslin** [reply]
The SYSREGS being set in the mmc_api,c allow MSSIO switching, it swaps the SD card settings to eMMC. This is being updated to use a function in the MPFS HAL, you could try calling [this](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/blob/main/driver-examples/mss/mss-mmc/mpfs-emmc-sd-write-read/src/application/hart1/u54_1.c#L232) before the MMC init. Mainly because in the default Icicle Kit configuration SD card is set up as the default from power up. If you call the switching function it'll swap you from the default (SD) to eMMC. If you added in the SYSREG writes from the mmc_api this should have done the same thing but its worth a try with the latest HAL.

**Reply by adaneyn** [reply]
Hi fcuzzocrea
Did you solve this issues or not yet? 
I am observing similar issue on my custom design board with both HSS and mpfs-emmc-sd-write-read application.
My board doesn't have SD card only 2x eMMC on each side of the switch. But the selected components are the same used in the ICICLE kit.
In mpfs-emmc-sd-write-read I have disabled case statement in u54_1.c that the code defaults into eMMC. But I keep getting eMMC failed message on the UART. I have even printed on the screen the error mss_mmc_status message and it seem to be MSS_MMC_OP_COND_ERR . But there is no explanation of that that could be. I have probed with the scope the CMD IO. And I have observed 2 busts one seem to be the soft reset and the second I guessed it is the attempt to initialize the eMMC chip form the MPFS but the eMMC chip is dosn't seem to respond to that message. At the moment it is hard for me to figure out it is a software or a hardware issue. 
There two possible issues one is that my eMMC is not working correctly but I don't see how because there is nothing exceptional in the eMMC chip just few data lines and  power which are exact copy of the ICILE kit. And power seem to be present on the eMMC chip. 
The other option is that the first command sent by the MPFS is not correct and the eMMC chip is not responding to that command so it is stuck in the initialization loop. I don't have any ICICLE kit to cross check but I have tested the mpfs-emmc-sd-write-read on Aries and it worked fine. The trouble is that the Aries has a diffrent eMMC chip than the ICICLE kit. The eMMC.
The other piece of trouble is that the eMMC protocol is based on JEDEC standard No. JESD84-B51 which is not publicly available. Normally as users we shouldn't have to dig so deep in the code or data to figure out what is the point of failure. But from this conversation it appears that the code has its issues too. All I want to know is if my problem is software related or hardware.

**Reply by fcuzzocrea** [reply]
At the end I got it working, but I took the whole code to initialize emmc from HSS. I would suggest to try running HSS on your custom board to cross check :)

Also my issue was also in the design. I was disabling muxing but did not reflashed the whole fpga design, just using the new XML on the bsp.

So what worked in my case was flashing fpga design and using hss code to initialize emmc

**Reply by adaneyn** [reply]
Thank you for sharing your experience.
What was the speed mode you have set in the MSS? I found that they have set it to HS400_ES in the ICICLE kit reference design.
I don't think that is correct for Scandisk DINBGG4-8G chip. 
Also have you modified any of the following parameters in mmc_init_emmc function?

        g_mmc.clk_rate = MSS_MMC_CLOCK_50MHZ;
        g_mmc.card_type = MSS_MMC_CARD_TYPE_MMC;
        g_mmc.bus_speed_mode = MSS_MMC_MODE_SDR;
        g_mmc.data_bus_width = MSS_MMC_DATA_WIDTH_8BIT;
        g_mmc.bus_voltage = MSS_MMC_1_8V_BUS_VOLTAGE;

**Reply by adaneyn** [reply]
Since my last message I made some progress but I still no totally sorted yet. I have found what was causing the eMMC chip to not respond.

I had the power supply of the core voltage not switched on.  I am using p-mos to enable the core voltage and I have tuned it off by mistake. 

However the chip is still failing at initialization. 

The function  ret_status = MSS_MMC_init(&g_mmc); retuns an MSS_MMC_INIT_FAILURE flag, 

I have used the debugger to follow find the exact point of failure and I found inside the mss_mmc.c there was a function ret_status = MSS_MMC_single_block_read(READ_SEND_EXT_CSD, csd_reg); 

Inside the fiction MSS_MMC_single_block_read the  isr_errors is returning errors when reading SRS12 register in line 826  isr_errors = MMC->SRS12;

the returned value is 0x0118000 which corresponds to SRS12_ERROR_INTERRUPT and SRS12_COMMAND_TIMEOUT_ERROR and SRS12_DATA_TIMEOUT_ERROR.

Any one has a clue of what could cause those flags to be asserted ?

**Reply by padmaraob** [reply]
@adaneyn

Please refer latest the mmc example:
https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/tree/main/driver-examples/mss/mss-mmc/mpfs-emmc-sd-write-read

If HS400_ES set in the ICICLE kit reference design means bank4 voltage is set for 1.8v,
Can you check in your design and board what is the bank4 voltage whether it is 1.8v or 3.3v?

also you can refer the icicle kit jumper settings for bank4
https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/boards/mpfs-icicle-kit-es/icicle-kit-user-guide/icicle-kit-user-guide.md#default-jumper-settings 

If you are getting SRS12_ERROR_INTERRUPT, SRS12_COMMAND_TIMEOUT_ERROR and SRS12_DATA_TIMEOUT_ERROR means the eMMC device is not responding? that may be because of incorrect clock or voltage supply.

**Reply by adaneyn** [reply]
Hi padmaraob
Thank you for trying to help me. 
I have used the latest BM design and my voltage on bank4 is set 1v8 in both hardware and MSS.
For the moment I am excluding the option that the eMMC device is not responding at all but I believe that it is the SRS12 interrogation that didn't end well. All the instructions in MSS_MMC_single_block_read are successful until it reaches line  line 826 isr_errors = MMC->SRS12;
The  if ((SRS12_ERROR_STATUS_MASK & isr_errors) == MMC_CLEAR) comparison is false which causes the MSS_MMC_init to abort and ret_status = MSS_MMC_ERR_INTERRUPT;
I personally believe that the MSS interruption is not working and causes the software to fail.

**Reply by padmaraob** [reply]
Hi adaneyn,

The MMC->SRS12 is used for read the status of host controller whether the cmd/data transfer go well or not.

In Your case, I think the command path is clear but not data path because when you are sending to read the data from eMMC device then the host controller status register(SRS12) is filled with errors.  
Can you please the clock frequency of MMC(400Khz or 50MHz for SDR mode) ? also check pull-up register value on the eMMC data path and compare with the icicle kit or sample eMMC device schematic?

**Reply by adaneyn** [reply]
Hello padmaraob
Changing the clock ad mode did help.
The register pulling and comparison with ICICLE kit is ongoing, 
You may be right about the data path but I am still thinking it is the software that has issues driving the eMMC chip.
I have probed the D0 and CMD and the both host and device seem to be chatting on those buses during the initialization stage before reaching the point of failure in the code. 
I have added a screen capture of the scope probing on D0 in yellow and CMD in blue and it looks to me like a sequence of command and response with data. I wouldn't expect to see any traffic on the remaining data lanes util the device is successfully initialized. 
After further investigation I have found that the real point of failure was CMD8 line of code:
  MMC->SRS03 = (uint32_t)((MMC_CMD_8_SEND_EXT_CSD << MMC_SRS03_COMMAND_SHIFT) | srs03_data);
That line seems to interrogate the chip EXT_CSD and if my understanding is correct the device should respond with the register values on the data bus and it is 512KB worth of data. But the device doesn't return any data to the host when executing line of code above.
So I don't find it logical that the D0 datapath works at the initial stage but not when executing that line of code with CMD8 command in it. 

![image](https://user-images.githubusercontent.com/99051309/221422108-b88380ea-61c9-43d4-b979-f16dbaae37ea.png)

**Reply by padmaraob** [reply]
Hi adaneyn,
The eMMC is set for the data width 1-bit, clock frequency 400Khz to read the EXT_CSD register first time with CMD8, the data lines are high when not used. The D0 is high when the host is sending the command 8 (with response R1) and D0 should toggle high/low when receiving data from the eMMC device.

The D0 is only low when the host is sending the command with busy response R1b and waiting to become high but this is not for CMD8.

**Comment by adaneyn** [comment]
Hello padmaraob
Thank you for your clarification.
I am just not sure from your answer if you are suggesting that the transition I am observing on d0 is from the host or device?

I can see two more transitions on D0 which are occurring at the following location in mss_mmc.c :
 /* Set MMC data bus width 1-bit */
                response_status = cif_send_cmd( MMC_DW_CSD |(MSS_MMC_DATA_WIDTH_1BIT << SHIFT_8BIT),
                                                MMC_CMD_6_SWITCH,
                                                MSS_MMC_RESPONSE_R1B);


 /* Set Phy delay for select MMC mode */
                    ret_status = phy_training_mmc(MSS_MMC_PHY_DELAY_INPUT_MMC_LEGACY, MSS_MMC_CLOCK_400KHZ);
I am putting the two corresponding waveforms
![image](https://user-images.githubusercontent.com/99051309/221664797-da88d535-d3eb-43aa-be8b-1fdc7da65b01.png)

![image](https://user-images.githubusercontent.com/99051309/221664880-fef0bc46-9a04-431f-a89f-de852238afff.png)

**Reply by padmaraob** [reply]
Hi adaneyn,
The wave-forms are Ok.
Can you please check the return status of phy_training_mmc() function?

If you get ret_status of phy_training_mmc() is "MSS_MMC_TRANSFER_SUCCESS" then the command and data path is clear for reading the data through the CMD8 or single/multi block command.
ret_status = phy_training_mmc(MSS_MMC_PHY_DELAY_INPUT_MMC_LEGACY, MSS_MMC_CLOCK_400KHZ);

**Comment by adaneyn** [comment]
Hello padmaraob
ret_status = phy_training_mmc(MSS_MMC_PHY_DELAY_INPUT_MMC_LEGACY, MSS_MMC_CLOCK_400KHZ); returns MSS_MMC_TRANSFER_SUCCESS and it is the waveform shown below.
but in MSS_MMC_single_block_read()  
 MMC->SRS03 = (uint32_t)((MMC_CMD_8_SEND_EXT_CSD << MMC_SRS03_COMMAND_SHIFT) | srs03_data);
doesn't return any data on D0. 


![image](https://user-images.githubusercontent.com/99051309/222028634-56deeb63-2154-4de1-9849-854381beb009.png)

**Reply by padmaraob** [reply]
Hi Adaneyn,

The single block read(CMD17) is used in the phy training to read data from sector(block) '0'. 

after phy_training_mmc() success, Can you try to write and read data on/from sector 0 and compare both write and read data is same or not, or any error? 

/* Set Phy delay for select MMC mode */
ret_status = phy_training_mmc(MSS_MMC_PHY_DELAY_INPUT_MMC_LEGACY, MSS_MMC_CLOCK_400KHZ);
if (ret_status == MSS_MMC_TRANSFER_SUCCESS)
{
	uint32_t loop_count;
	uint8_t tx_data_buffer[512] = {0u};
	uint8_t rx_data_buffer[512] = {0u};
    g_mmc_init_complete = MMC_SET;
	for (loop_count = 0; loop_count < 512; loop_count++)
	{
		tx_data_buffer[loop_count] = 0x45 + loop_count;
	}
    
	ret_status = MSS_MMC_single_block_write(tx_data_buffer, 0);
	if (MSS_MMC_TRANSFER_SUCCESS == ret_status)
	{
		ret_status = MSS_MMC_single_block_read(0, rx_data_buffer);
	}

**Comment by adaneyn** [comment]
Hello padmaraob
The two functions MSS_MMC_single_block_write and MSS_MMC_single_block_read are not working
The ret_status retuns MSS_MMC_NOT_INITIALISED

**Reply by padmaraob** [reply]
Hi adaneyn,
Missed one line of code in my above sample code.
“g_mmc_init_complete = MMC_SET;”
Updated now in previous comment

**Comment by adaneyn** [comment]
Yes it worked! I am very grateful to you. 
ret_status returns MSS_MMC_TRANSFER_SUCCESS for both functions.
So why it is working here and not in MSS_MMC_init function?

![image](https://user-images.githubusercontent.com/99051309/222195804-30f766ec-957d-49b3-ae5e-b6c16e1e2203.png)

**Reply by adaneyn** [reply]
Hello padmaraob
Would you think that the EXT_CSD register got corrupted somehow in the chip?
Example due to handling or faulty batch? I personally find it hard to believe but could explain why EXT_CSD  doesn't return anything.
Maybe I could interrogate another register such as CSD with CMD27 and see if I get any return? 
For information, I am using two SDINBDG4-8G-ZA with a TS3A27518EIPWRQ1 switches on my board.  One of SDINBDG4-8G-ZA is placed instead of the SD card in the ICICLE kit. Yet both chips do not respond to CMD8.

**Reply by padmaraob** [reply]
Hi adaneyn,
don't know exactly, why it is not reading the EXT_CSD.
anyway the CSD read is working fine when reading with cmd9. so no need to program with cmd27.

**Reply by adaneyn** [reply]
Yes I think at this stage we are both running out of clues. I will contact the support team for the Scandisk device. Maybe change the chip.
Is there a trick to skip EXT_CSD and initialize the chip?

**Reply by adaneyn** [reply]
Hi Padmaraob
Eventually after replacing the device it worked!
Something must have gone terribly wrong to have the two chips failing in the same way. 
But it was also good to acquire some deeper knowledge about eMMC. 
Thank you for your help you have been very supportive and contributed greatly to finding the problem.
I am very grateful to you and Sricker.


![image](https://user-images.githubusercontent.com/99051309/223883359-da0bd0cb-648f-455f-97a6-9c8ea30fca3a.png)
